### ISSUE TITLE
OIndexTxAware* search for records, changed inside transaction, dont follow index definition

### ISSUE THREAD

**Issue by vantonov1** [orientechnologies/orientdb#6537]
v 2.2.3

OIndexTxAware dont use underlying delegate (real) index to search inside transaction changes, only simple indexChanges.getChangesPerKey(key) is used

As a result, it is impossible to search for substrings, *, use collated search etc

**Comment by taburet** [comment]
Hi @vantonov1! Failed to reproduce the problem with the following test case:

```java
import com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx;
import com.orientechnologies.orient.core.metadata.schema.OClass;
import com.orientechnologies.orient.core.metadata.schema.OType;
import com.orientechnologies.orient.core.record.impl.ODocument;
import com.orientechnologies.orient.core.sql.OCommandSQL;
import com.orientechnologies.orient.core.sql.query.OResultSet;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * @author Sergey Sitnikov
 */
public class TxAwareIndexTest {

  private ODatabaseDocumentTx db;

  @Before
  public void before() {
    db = new ODatabaseDocumentTx("memory:TxAwareIndexTest");
    db.create();
    db.getMetadata().getSchema().createClass("user").createProperty("name", OType.STRING).createIndex(OClass.INDEX_TYPE.UNIQUE);
    db.newInstance("user").field("name", "abc").save();
    db.newInstance("user").field("name", "aby").save();
    db.newInstance("user").field("name", "abz").save();
  }

  @After
  public void after() {
    db.drop();
  }

  @Test
  public void testSubstrings() {
    db.begin();
    db.command(new OCommandSQL("update user set name='abd' where name='aby'")).execute();
    OResultSet<ODocument> r = db.command(new OCommandSQL("select * from user where name like '%b%' order by name")).execute();
    assertEquals(3, r.size());
    assertEquals("abc", r.get(0).field("name"));
    assertEquals("abd", r.get(1).field("name"));
    assertEquals("abz", r.get(2).field("name"));

    r = db.command(new OCommandSQL("select * from user where name > 'abc' order by name")).execute();
    assertEquals(2, r.size());
    assertEquals("abd", r.get(0).field("name"));
    assertEquals("abz", r.get(1).field("name"));
    db.commit();
  }

}
```

Also, inspected the source code of `OIndexTxAware*` classes, seems like they properly combine information from the transaction with the information in the underlying index. Maybe I just missed something. Could please modify the test case above to make it fail or describe the steps to reproduce more precisely or maybe point the method of `OIndexTxAware*` containing the erroneous code?

**Comment by vantonov1** [comment]
Hi!

I checked, your test doesnt call for calculateTxIndexEntry() - thats the only place both OIndexTxAwareMultiValue and OIndexTxAwareOneValue call getChangesPerKey(). It seems not every select use iterateEntries()

Really, I couldnt restore details right now, I'll do it a little bit later, sorry

**Comment by vantonov1** [comment]
Ha, thats simple. It seems that only IN uses iterateEntries(), and at least it doesnt uses collation from index. So, here is the test

`        ODatabaseDocumentTx db = new ODatabaseDocumentTx("memory:TxAwareIndexTest");
        db.create();
        final OClass user = db.getMetadata().getSchema().createClass("user");
        user.createProperty("name", OType.STRING)/*.createIndex(OClass.INDEX_TYPE.UNIQUE)*/;
        final ODocument metadata = new ODocument();
        metadata.field("COLLATE", "ci");
        user.createIndex("name", "unique_hash_index", null, metadata, new String[]{"name"});
        db.newInstance("user").field("name", "abc").save();
        db.newInstance("user").field("name", "aby").save();
        db.newInstance("user").field("name", "abz").save();
        db.begin();
        db.command(new OCommandSQL("update user set name='abd' where name='aby'")).execute();
        OResultSet<ODocument> r = db.command(new OCommandSQL("select * from user where name in ['Abd'] order by name")).execute();
        Assert.assertEquals(1, r.size());
`

Sorry for imprecise subject

**Comment by vantonov1** [comment]
BTW, how your insert a code with perfect formatting? I couldnt

**Comment by taburet** [comment]
Thanks for the test case. Formatting is simple as:

    ```java
    // java code here
    ```

**Comment by taburet** [comment]
Fixed in 2.2.x and develop branches. Now indexes respect collations during transaction.
